### DISCUSSION TITLE
ConfigurationProperties does not resolve from environment variable(s) in some cases

### DISCUSSION THREAD

**Discussion by MikiLoz92** [micronaut-projects/micronaut-core#4362] [Q&A] [answered]
I have encountered a really particular case in which I'm unable to resolve a ConfigurationProperty out of an environment variable.

I have the following interface created:
```kotlin
@ConfigurationProperties("elastic")
interface ElasticConfiguration {
    val cloudId: String
    val username: String
    val password: String
    val indexPrefix: String
}
```

Then I inject this configuration in another class:
```kotlin
@Singleton
class ElasticsearchDocumentStorer(
        private val elasticConfiguration: ElasticConfiguration
) : CoroutineScope {

    private val indexPrefix = elasticConfiguration.indexPrefix

    suspend fun storeDocument(index: String, document: JsonObject) {
        val indexName = "${indexPrefix}-$index"
        ...
    }
}
```
This works OK, however if I change the `storeDocument` method's contents to this:
```kotlin
val indexName = "${elasticConfiguration.indexPrefix}-$index"
```
micronaut throws the following error at runtime, when accessing that method:
```
[astra-gateway] io.micronaut.core.value.PropertyNotFoundException: No property found for name [elastic.index-prefix] and type [java.lang.String]
[astra-gateway]         at io.micronaut.runtime.context.env.ConfigurationIntroductionAdvice.lambda$intercept$1(ConfigurationIntroductionAdvice.java:109)
[astra-gateway]         at java.util.Optional.orElseThrow(Optional.java:290)
[astra-gateway]         at io.micronaut.runtime.context.env.ConfigurationIntroductionAdvice.intercept(ConfigurationIntroductionAdvice.java:109)
[astra-gateway]         at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:82)
[astra-gateway]         at eos.connectivity.backend.astra.application.configuration.ElasticConfiguration$Intercepted.getIndexPrefix(Unknown Source)
[astra-gateway]         at eos.connectivity.backend.astra.application.infrastructure.persistence.ElasticsearchDocumentStorer.storeDocuments(ElasticsearchDocumentStorer.kt:62)
[astra-gateway]         at eos.connectivity.backend.astra.session.store.infrastructure.persistence.ElasticsearchFleetSessionStatusRepository.store(ElasticsearchFleetSessionStatusRepository.kt:26)
[astra-gateway]         at eos.connectivity.backend.astra.session.store.FleetSessionStatusUpdater.invoke(FleetSessionStatusUpdater.kt:10)
[astra-gateway]         at eos.connectivity.backend.astra.session.store.infrastructure.subscribers.StoreFleetSessionStatusOnFleetSessionStatusUpdatedEventSubscriber$invoke$1.invokeSuspend(StoreFleetSessionStatusOnFleetSessionStatusUpdatedEventSubscriber.kt:22)
[astra-gateway]         at eos.connectivity.backend.astra.session.store.infrastructure.subscribers.StoreFleetSessionStatusOnFleetSessionStatusUpdatedEventSubscriber$invoke$1.invoke(StoreFleetSessionStatusOnFleetSessionStatusUpdatedEventSubscriber.kt)
[astra-gateway]         at eos.connectivity.backend.astra.application.ExtensionsKt$fireSafeCoroutine$1.invokeSuspend(Extensions.kt:14)
[astra-gateway]         at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[astra-gateway]         at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
[astra-gateway]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[astra-gateway]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[astra-gateway]         at java.lang.Thread.run(Thread.java:748)

```

The property should be resolving from an environment variable called ELASTIC_INDEX_PREFIX, which at the moment of executing the method is there (I'm printing all env variables and I can see it right before). If I put the property in the application.yml file (as `elastic.index-prefix`), it resolves just fine and there are no errors.

This is happening in Micronaut 2.1.1, and was happening in Micronaut 1.3.5 as well, so I tried to update to see I could fix it, with not luck.

Unfortunately, I have been unable to provide a reproducer for this issue; it is only happening under my current project & setup, which uses Kubernetes.

If I can be of any help, I am available for a live screen sharing of this issue or private sharing of the code.

**Comment by alvarosanchez** [comment]
Can you [enable the `/env` endpoint](https://docs.micronaut.io/latest/guide/index.html#environmentEndpoint) and check it to see what configuration properties are resolved?

**Comment by MikiLoz92** [comment]
Sure, I'll try to do that today. Thanks for the tip.

**Comment by MikiLoz92** [comment]
I'm not able to start up the Micronaut HTTP server. It says everytime:
```
[INFO ] 2020-10-22 12:39:51.168 [main] Micronaut: 150 - No embedded container found. Running as CLI application
```
I'm running the fatjar created by shadowJar and I've tried every possible fix related to that error that I've found online.

It may be because I'm using Vert.x to provide an HTTP server...? I've disabled to Vert.x HTTP server itself, but I cannot make Micronaut start its own server whatsoever...

**Comment by MikiLoz92** [comment]
Saying this because AFAIK Vert.x also uses Netty for its HTTP server implementation...

**Comment by graemerocher** [comment]
Sounds like you have a pretty complex setup with lots of issues not related to Micronaut itself but more to your project. I suggest to progress this forward you provide an example that reproduces the issue.

**Comment by alvarosanchez** [comment]
@MikiLoz92 you are likely missing Micronaut HTTP server dependencies. Compare your build with a fresh one generated from https://launch.micronaut.io and try to see what are you missing

**Comment by MikiLoz92** [comment]
I have managed to procure a reproducer for this issue. It only seems to happen when I put Vert.x as a dependency and use its timer.

https://github.com/MikiLoz92/micronaut-configurationpropertiesbug-reproducer

To reproduce, run the main function and wait 5 seconds for the timer to trigger. You should see an error. If you uncomment the same property on the application.yml file, the error is gone.

I am printing the ENV variables just before the error, and I see the variable that this property is referring to just there, so that's weird.

**Comment by MikiLoz92** [comment]
@alvarosanchez I did have the Micronaut HTTP server dependencies on my project, but I guess that's a separate issue...

**Comment by graemerocher** [comment]
You are using `ApplicationContext.run()` instead of `Micronaut.run()`, the former only starts the context no the server

**Comment by MikiLoz92** [comment]
The same is happening with Micronaut.run(), could you take another look at it @graemerocher ??

Also, could you direct me to the place in the documentation that states that ConfigurationProperties are only usable from environment variables if and only when the Micronaut server is started?

**Comment by MikiLoz92** [comment, marked as answer]
stacktrace w/ `Micronaut.run()`:
```
16:04:20.311 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [gcp, cloud]
16:04:26.861 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 6787ms. Server Running: http://localhost:8080
PATH=/Users/miki/Downloads/google-cloud-sdk/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
SHELL=/bin/bash
USER=miki
TMPDIR=/var/folders/1y/q__rck6569n6v3z67hlr892h0000gn/T/
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.fkdUchCobI/Listeners
XPC_FLAGS=0x0
VERSIONER_PYTHON_VERSION=2.7
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x2
LOGNAME=miki
LC_CTYPE=en_GB.UTF-8
PWD=/Users/miki/configurationproperties-bug-reproducer
XPC_SERVICE_NAME=com.jetbrains.intellij.15148
ELASTIC_INDEX_PREFIX=beta
JAVA_MAIN_CLASS_34760=configurationproperties.bug.reproducer.ApplicationKt
HOME=/Users/miki

beta-indexname
Oct 22, 2020 4:04:32 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
io.micronaut.core.value.PropertyNotFoundException: No property found for name [elastic.index-prefix] and type [java.lang.String]
	at io.micronaut.runtime.context.env.ConfigurationIntroductionAdvice.lambda$intercept$1(ConfigurationIntroductionAdvice.java:109)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at io.micronaut.runtime.context.env.ConfigurationIntroductionAdvice.intercept(ConfigurationIntroductionAdvice.java:109)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:82)
	at configurationproperties.bug.reproducer.ElasticConfiguration$Intercepted.getIndexPrefix(Unknown Source)
	at configurationproperties.bug.reproducer.ElasticsearchDocumentStorer.storeDocument(ElasticsearchDocumentStorer.kt:15)
	at configurationproperties.bug.reproducer.FleetSessionStatusPeriodicNotifier$start$1.handle(FleetSessionStatusPeriodicNotifier.kt:15)
	at configurationproperties.bug.reproducer.FleetSessionStatusPeriodicNotifier$start$1.handle(FleetSessionStatusPeriodicNotifier.kt:8)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:919)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:887)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:221)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:913)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:176)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

16:04:33.242 [Thread-1] INFO  io.micronaut.runtime.Micronaut - Embedded Application shutting down

Process finished with exit code 130 (interrupted by signal 2: SIGINT)
```

**Reply by graemerocher** [reply]
The issue is because you are starting the Vertx service before Micronaut has finished initialization. The issue can be resolved with a startup listener:

```kotlin
import io.micronaut.context.event.StartupEvent
import io.micronaut.runtime.Micronaut
import io.micronaut.runtime.event.annotation.EventListener
import javax.inject.Singleton

fun main(args: Array<String>)  {
	Micronaut.run()
}

@Singleton
open class Application(private val fleetSessionStatusPeriodicNotifier: FleetSessionStatusPeriodicNotifier) {
	@EventListener
	fun onStart(event : StartupEvent) {
		fleetSessionStatusPeriodicNotifier.start()
	}
}
```

**Reply by MikiLoz92** [reply]
OK, thank you so much for the tip. I've tried it in my project and it's definitely this, because now it's working fine.

Just to make everything clear... this is happening because there is a weird interaction with Vert.x or just because I wasn't waiting for the IoC container to be ready (I'm not using Micronaut's server either and now that I've done it this way it's working)?

**Reply by jameskleeh** [reply]
> just because I wasn't waiting for the IoC container to be ready

@MikiLoz92  This

**Comment by MikiLoz92** [comment]
Sorry to bump this again, but has something been changed on 2.3.0 regarding the way @ConfigurationProperties resolve from env variables? 

Because one of our applications that was working fine on 2.2.2 (after implementing the changes we discussed in this thread), is now failing to start on 2.3.0 because is fails to resolve from env vars. Same setup, same everything, just by changing the version to 2.3.0 it fails.

**Reply by jameskleeh** [reply]
@MikiLoz92 Known issue with 2.3.0, its resolved for 2.3.1 which will be released today

**Reply by MikiLoz92** [reply]
great! thank you 👍
